lib/modem_key_mgmt/modem_key_mgmt.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <at_cmd.h>
+#include <at_cmd_parser/at_cmd_parser.h>
+#include <at_cmd_parser/at_params.h>
+#include <bsd_limits.h>
+#include <modem_key_mgmt.h>
+#include <stdio.h>
+#include <string.h>
+#include <zephyr.h>
+
+#define MODEM_KEY_MGMT_OP_LS "AT%CMNG=1"
+#define MODEM_KEY_MGMT_OP_RD "AT%CMNG=2"
+#define MODEM_KEY_MGMT_OP_WR "AT%CMNG=0"
+#define MODEM_KEY_MGMT_OP_RM "AT%CMNG=3"
+
+#define AT_CMNG_PARAMS_COUNT 5
+#define AT_CMNG_CONTENT_INDEX 4
+
+static char scratch_buf[4096];
+static char at_cmee_strings[2][10] = {"AT+CMEE=0", "AT+CMEE=1"};
+
+static int cmee_active(void)
+{
+	int result;
+	char response[sizeof("+CMEE: X\r\n") + 1];
+
+	memset(response, 0, sizeof(response));
+
+	result = at_cmd_write("AT+CMEE?", response, sizeof(response), NULL);
+	if ((result == 0) && strlen(response) > 0) {
+		if (strchr(&response[6], '1')) {
+			return 1;
+		} else {
+			return 0;
+		}
+	}
+
+	return -EIO;
+}
+
+static int cmee_set(bool enable)
+{
+	return at_cmd_write(at_cmee_strings[(int)enable], NULL, 0, NULL);
+}
+
+static int translate_error(int err)
+{
+	switch (err) {
+	case 513:
+		return -ENOENT;
+	case 514:
+		return -EPERM;
+	case 515:
+		return -ENOMEM;
+	case 518:
+		return -EACCES;
+	default:
+		return -EFAULT;
+	}
+}
+
+static int write_at_cmd_with_cme_enabled(char *cmd, char *buf, size_t buf_len,
+					 enum at_cmd_state *state)
+{
+	int retval;
+	int cmee_was_active = cmee_active();
+
+	if (cmee_was_active < 0) {
+		return -EFAULT;
+	}
+
+	if (!cmee_was_active) {
+		if (cmee_set(true)) {
+			return -EIO;
+		}
+	}
+
+	retval = at_cmd_write(cmd, buf, buf_len, state);
+
+	if (cmee_was_active) {
+		if (cmee_set(false)) {
+			return -EIO;
+		}
+	}
+
+	return retval;
+}
+
+int modem_key_mgmt_write(nrf_sec_tag_t sec_tag,
+			 enum modem_key_mgnt_cred_type_t cred_type,
+			 const u8_t *buffer, u16_t buffer_len)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	if ((buffer == NULL) || (buffer_len == 0)) {
+		return -EINVAL;
+	}
+
+	written = snprintf(scratch_buf, sizeof(scratch_buf), "%s,%u,%u,\"",
+			   MODEM_KEY_MGMT_OP_WR, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+	if ((written < 0) || (written >= sizeof(scratch_buf))) {
```

Comment:
`written` can't be larger than the size of the input buffer, it can be at most `sizeof(scratch_buf) - 1`, so checking
```
written >= sizeof(scratch_buf))
```
is not correct.

Reply:
Can you elaborate? Checking for `written >= sizeof(scratch_buf))` does exactly that - verifies that `written` value is at most `sizeof(scratch_buf) - 1`.

Reply:
ok, let me rephrase: it's not that it's not correct but I didn't see why are we checking for that.
I thought that `snprintf` couldn't return a value larger than the size of the input buffer, but  I have re-read the documentation and I think it's correct now to check that the return value is both less than the input size and positive.

Reply:
Yes, one could expect that `snprintf` would simply return an error in case the buffer was too small, apparently, this function works differently.